### ISSUE TITLE
Memory Leak since 2.1.3

### ISSUE THREAD

**Issue by jgordijn** [zio/zio#9002]
Since a new release of our product with the latest ZIO (2.1.5) we are experiencing memory leaks. We do not use the `flatMapParUnordered`. I saw that in 2.1.3 there is a fiber worker pool for parallel processing. This pool seems to be the size of `ZIO.Parallelism` which I think is `None`, by default. 

Does this translate into unlimited Fibers?

We use ZIO-kafka's `partitionedStream.flatMapPar(Int.MaxValue)` for processing partitions of a topic, and in one place `ZIO  .foreachPar(list)(...)` for handling a list. Apart from that, we use Tapir (with http4s) to process HTTP requests.

Is there any metric I can supply to debug this issue?

**Comment by jgordijn** [comment]
This is what I got from a heap dump analyzed by the Eclipse Memory Analyzer:
```
282.487 instances of zio.internal.FiberRuntime, loaded by jdk.internal.loader.ClassLoaders$AppClassLoader @ 0x91300ad0 occupy 508.972.680 (58,10%) bytes.

Most of these instances are referenced from one instance of java.util.concurrent.RunnableScheduledFuture[], loaded by <system class loader>, which occupies 1.104 (0,00%) bytes. The instance is referenced by java.lang.Thread @ 0xa2225418 zio-timer-1, loaded by <system class loader>.

The thread java.lang.Thread @ 0xa2225418 zio-timer-1 keeps local variables with total size 496 (0,00%) bytes.

Thread java.lang.Thread @ 0xa2225418 zio-timer-1 has a local variable or reference to java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue @ 0xa22254a8 which is on the shortest path to java.util.concurrent.RunnableScheduledFuture[271] @ 0xa04d9620. The thread java.lang.Thread @ 0xa2225418 zio-timer-1 keeps local variables with total size 496 (0,00%) bytes.
```

With the now downgraded version I don't get any mention of the number FiberRuntime instances

**Comment by kyri-petrou** [comment]
Hi @jgordijn, there was in a similar issue in v2.1.3 with `mapZIOParUnordered` but it should have been fixed AFAIK. I think as an immediate remedy you should try and limit the parallelism level in `flatMapPar` (instead of using `Int.MaxValue`). It's highly dependent on the number of CPUs that the pod uses and the workflow executed in parallel, but I think setting a value of ~ 4-8 * nCPU probably makes most sense. If the workflows are fully async you could potentially increase it more, but definitely try and keep it relatively low.

In the meantime, @eyalfa would you be able to look into what might be causing this memory leak?

**Comment by narma** [comment]
@kyri-petrou issue with `mapZIOParUnordered` currently fixed only in latest snapshot versions, release with fix not published yet.

**Comment by narma** [comment]
Short reproducer:

```scala
ZStream
        .iterate(1)(_ + 1)
        .flatMapPar(Int.MaxValue) { i =>
          ZStream.fromZIO(ZIO.when(i % 1000 == 0)(ZIO.logDebug(s"$i")).delay(10.milliseconds))
        }
        .runDrain
```

So flatMapPar at least with Int.MaxValue parallelism is leak currently

**Comment by narma** [comment]
Just added tests for memory leaks on these operators in PR https://github.com/zio/zio/pull/9003

**Comment by jgordijn** [comment]
Thanks for confirming @narma . Is `foreachPar` relying on the same mechanism?

I wonder if the new approach with the Fiber pool is also memory tested? Isn't it considered an antipattern to create a pool of Fibers? Just like it is with Virtual Threads? I thought Fibers were lightweight and meant to be short-lived?

**Comment by myazinn** [comment]
@narma your case doesn't seem like a memory leak. The producer in that snippet is much faster than the consumer, and it creates _a lot_ of fibers before any of them finishe and ready to pick up the next task. Given that parallelism is requested to be effectively unbounded, that seems like an expected behaviour.
I tried to run it and throttle the producer to 10k el/s, and it seems to work fine. No major increase in memory consumption in almost 30 minutes.
```scala
    ZStream
      .iterate(1)(_ + 1)
      .rechunk(1000)
      .throttleShape(10000, 1.second)(_.size)
      .flatMapPar(Int.MaxValue) { i =>
        ZStream.fromZIO(ZIO.when(i % 1000 == 0)(ZIO.debug(s"$i")).delay(10.milliseconds))
      }
      .runDrain
```
<img width="586" alt="image" src="https://github.com/zio/zio/assets/28982711/17d380c4-9780-452b-9d89-1f040136b6ab"> \
I guess if we run the initial snippet with sufficient resources (hundreds of CPUs and terabytes of RAM), it will run ok and consume near-constant amount of resources.

I agree with @jgordijn that the issue seems to be with underlying fiber pool (though there's no explicit real pool). It can only grow over time and never releases unused fibers. E.g. in the following example a lot of fibers get created immediately and then never released until the ZStream is over.
```scala
  override def run: ZIO[Any, Any, Any] = {
    val s1 = ZStream.iterate(1)(_ + 1).take(1000000)
    val s2 = ZStream.repeatZIO(ZIO.succeed(1000).delay(1.second))

    (s1 ++ s2)
      .flatMapPar(Int.MaxValue) { i =>
        ZStream.fromZIO(ZIO.when(i % 1000 == 0)(ZIO.debug(s"$i")).delay(10.milliseconds))
      }
      .runDrain
  }
```
<img width="586" alt="image" src="https://github.com/zio/zio/assets/28982711/db41cc85-2f05-47a5-98b8-31cb11245607"> \
I imagine it could easily lead to sort of memory leak, since despite the internal "best effort" new fibers might be created from time to time, thus consuming more and more resources.

**Comment by jgordijn** [comment]
To be clear, the the stream from Kafka will create a substream per partition. So the amount of elements there are limited. Also, we use this in many services, but only this service has this issue. 

This service does more http and we use the ZIO.forEachPar on results from the database. Not many services use this ZIO.forEachPar.

**Comment by kyri-petrou** [comment]
@jgordijn I'm fairly convinced that the issue is with `partitionedStream.flatMapPar(Int.MaxValue)`. I somehow doubt `ZIO.foreachPar` is the culprit here but I could be wrong so we can't exclude that possibility.

Until we get to the bottom of it, could you try the following things and let me know if any of these work?
1. Try `partitionedStream.flatMapPar(32)` (or some other low parallelism value) instead of `Int.MaxValue`
2. Try using the `ZIO.withParallelism` to wrap calls to `ZIO.foreachPar` e.g.,

```scala
ZIO.withParallelism(128) {
  ZIO.foreachPar(???)(???)
}
```

My gut feeling is that the culprit is (1) but let's see if (2) makes any difference as well

**Comment by jgordijn** [comment]
@kyri-petrou yes, I will try this one by one. It will take some time to see the results. I'll report back

**Comment by jgordijn** [comment]
I tried `partitionedStream.flatMapPar(64)`, but I still see the memory going up. So that does not solve it. Which would make sense, as there is not much parallelism there anyways. 1 per partition. I'll try the ZIO.withParallelism

**Comment by kyri-petrou** [comment]
Thanks @jgordijn. Would you be able to put a small reproducer together by any chance? I have an idea on what might be causing it but it's not straightforward to fix and want to have something to test against

**Comment by kyri-petrou** [comment]
Also, can you double-check whether v2.1.6 that was released yesterday didn't solve the issue?

**Comment by jgordijn** [comment]
With nr 2 I still have the memory issue. I didn't try 2.1.6 yet

**Comment by jgordijn** [comment]
Version 2.1.6, still leads to memory leak.

**Comment by eyalfa** [comment]
@kyri-petrou can u please tldr this for me? is there a leaking stream operator or not?
I recall fixing one of them few weeks back, but it's quite possible the same pattern exists in other operators as well.

re. one of the comments about using a fibers pool being an anti-pattern, I came to this solution via profiling and benchmarking. as always with these things it's worth re-evaluating every once in a while (Kyrie and others have invested a lot into fibers performance lately).

**Comment by kyri-petrou** [comment]
@eyalfa A very TLDR is this:

> I agree with @jgordijn that the issue seems to be with underlying fiber pool (though there's no explicit real pool)

I run the snippet that @myazinn provide, did a memory snapshot, and there is an insanely high number of objects that are kept referenced by the ZStream due to the fiber pooling and therefore are not GCd until the ZStream is closed. Since ZStreams / ZChannels can be extremely long-lived (e.g., Kafka consumers), they need to be as stateless as humanly possible and not hold reference to anything other than what they're currently processing. Otherwise we end up with memory leaks.

Do you think it's worth trying to refactor the existing parallel implementations on ZStream or are we better off reverting the implementation to the one in pre-2.1.3?

**Comment by eugeny-stoyka** [comment]
Any workaround ZStream with iterable?

**Comment by eyalfa** [comment]
@kyri-petrou I'll poke into this tomorrow, I believe it's possible to refine the new implementation to 'loose' excessive fibers, alternatively forking may have become cheap enough to drop the pool approach.

btw, the way I understand it the issue is limited to the unbounded parallelism case. is that right? 
notice the implementation cannot just decide on a high enough number and cap the fibers, this is something the application must do (I think one of the tests demonstrates this)

**Comment by eyalfa** [comment]
@kyri-petrou 
while I accept @narma 's reproducer leaks, I suspect this is not the case with kafka's partitioned stream.
for two reason:
1. limited parallelism, after all the number of partitions is bounded
2. I suspect once a partition is 'opened' it'd only close when upstream completes which is basically when the pod is terminated, so sub-streams live forever, hence their handling fiber lives forever, both in the new and old implementations.

**Comment by eyalfa** [comment]
> Any workaround ZStream with iterable?
@eugeny-stoyka can u elaborate? what's your use-case?
in general the current workaround is to limit parallelism to a reasonable number (depends on your scenario).

just a hunch, but if your Iterable has finite size (most do), the 'unbounded' parallelism is effectively bounded by the Iterable size. notice this still behaves differently from the previous implementation as the fibers (currently) keeps lingering even when there's no further work.

**Comment by kyri-petrou** [comment]
> btw, the way I understand it the issue is limited to the unbounded parallelism case. is that right?

If bounded parallelism is set to a large enough value, it will also have the same issue, so we can't neglect it.

@eyalfa I thought about this a bit more, and I don't think the issue is with having a fiber pool per-se. The problem as I see it is the pool can only grow in size but can never shrink when its no longer needed. AFAICT this isn't an issue just with `flatMapPar` but to all the `Par` implementations.

I think we need a design that allows creating up to `n` fibers (based on the user defined parallelism) and discard them when we no longer need them (maybe when we try to poll on an empty queue?). WDUT?

**Comment by eyalfa** [comment]
I'm basically looking into implementing something like that atm, thing is it requires for some heuristics and perhaps more (optional) user input. i.e. dropping fibers on an empty poll (which is relatively easy) can lead to fork-kill loops where we kill fibers 'just before' we need them, so this leads to parameterized heuristics based on minimum number of fibers to keep, how long can we block on the buffer before dropping the fiber, not to mention how can this be implemented without allocating extra fibers (i.e. `ZIO#timeout`) or paying a performance penalty.

**Comment by eyalfa** [comment]
@kyri-petrou on second thought I don't think @narma 's reproducer is an issue and as @myazinn commented it behaves as expected, combining a fast producer, slow consumer and unbounded parallelism results with unbounded fibers - in both implementations.

@myazinn 'throttled' scenario is more concerning as it does indicate a regression in behavior, this basically supports your conclusion about the pool lacking the ability to shrink.

**Comment by eugeny-stoyka** [comment]
> > Any workaround ZStream with iterable?
> > @eugeny-stoyka can u elaborate? what's your use-case?
> > in general the current workaround is to limit parallelism to a reasonable number (depends on your scenario).
> 
> just a hunch, but if your Iterable has finite size (most do), the 'unbounded' parallelism is effectively bounded by the Iterable size. notice this still behaves differently from the previous implementation as the fibers (currently) keeps lingering even when there's no further work.

If you finite something it will slow down your computations, but it has leaks in memory anyway, as i see in my opinion

```scala
object MyZIOApp extends ZIOAppDefault {

  val fileIterator: Iterator[String] = Source.fromResource("ips.csv").getLines() //100 million ips
  val fileStreamIterator: ZStream[Any, Throwable, String] = ZStream.fromIterator(fileIterator)

  def zStreamIp: ZStream[Any, Throwable, IPv4] = {
    fileStreamIterator
        .map(_.split(","))
        .flatMap {
          case Array(fromIP, toIP, _*) =>
            ZStream.fromIterator(IP(fromIP) to IP(toIP))
        }
  }

  def run: ZIO[Any, Throwable, Unit] = {
    zStreamIp.mapZIOParUnordered(Int.MaxValue) {
      ip =>
        for {
          ipaddr <- InetAddress.byName(ip.toString)
          r <- ZIO.blocking(ipaddr.isReachable(10000))
          _ <- if(r) Console.printLine(ip) else ZIO.none
        } yield ()
    }.runDrain
  }
}
```

**Comment by kyri-petrou** [comment]
> If you finite something it will slow down your computations

@eugeny-stoyka just want to touch on this a bit. Do you mean that using a limit on parallelism will make things slower? If that's what you mean, then this is very incorrect.

Parallelism always incurs additional overhead. This is not just with ZIO, but with almost all systems. Depending on your workfow and the number of resources available to your system (CPU, memory), it is _always_ better for parallelism to be bounded, and most of the times, it should be much lower than what you think you need.

For more info, make sure to check out [this post](https://github.com/brettwooldridge/HikariCP/wiki/About-Pool-Sizing)(and the video!). While the technology is different (postgres), a lot of the fundamentals are applicable for all systems, including ZIO.

EDIT: There is a fair bit of literature online on thread pool sizing / limiting parallelism, but you can use this formula as a starting point `Number of Cores * (1 + Blocking Coefficient)` where the Blocking Coefficient is the wait-to-CPU time ratio ([source](https://www.linkedin.com/pulse/demystifying-java-thread-pool-sizing-insights-optimal-rathnayake-rm47c))

**Comment by jgordijn** [comment]
> Thanks @jgordijn. Would you be able to put a small reproducer together by any chance? I have an idea on what might be causing it but it's not straightforward to fix and want to have something to test against

I cannot reproduce it yet. I'm unsure where it comes from. We have many applications that process data from Kafka with the same setup without problems. So I suspect the issue is coming from the HTTP stack (tapir, http4s, netty).

**Comment by jgordijn** [comment]
> re. one of the comments about using a fibers pool being an anti-pattern, I came to this solution via profiling and benchmarking. as always with these things it's worth re-evaluating every once in a while (Kyrie and others have invested a lot into fibers performance lately).

I know for Java Virtual Thread it is deemed an anti-pattern. Not sure for Fibers, but I can imagine that the idea of Fibers is not to pool them. 

Did you also profile memory usage? Being faster is only 1 aspect. Shouldn't the solution be to improve the performance of Fibers, instead of pooling them.

**Comment by jgordijn** [comment]
> @kyri-petrou while I accept @narma 's reproducer leaks, I suspect this is not the case with kafka's partitioned stream. for two reason:
> 
> 1. limited parallelism, after all the number of partitions is bounded

Yes. I tested with a bounded parallelism with still the same effect indeed. 

> 2. I suspect once a partition is 'opened' it'd only close when upstream completes which is basically when the pod is terminated, so sub-streams live forever, hence their handling fiber lives forever, both in the new and old implementations.

Kafka has rebalancing, so it will also close streams when a new pod joins or leaves. This doesn't happen often in the case where we have memory leaks though.

**Comment by kyri-petrou** [comment]
@jgordijn any chance you could take a heap dump and identify the objects that are being leaked? And perhaps try to also check if you can find where those objects are being referenced from?

Also, are you able to reproduce this leak locally do you need to deploy changes in order to emulate it?

**Comment by eugeny-stoyka** [comment]
> > If you finite something it will slow down your computations
> 
> @eugeny-stoyka just want to touch on this a bit. Do you mean that using a limit on parallelism will make things slower? If that's what you mean, then this is very incorrect.
> 
> Parallelism always incurs additional overhead. This is not just with ZIO, but with almost all systems. Depending on your workfow and the number of resources available to your system (CPU, memory), it is _always_ better for parallelism to be bounded, and most of the times, it should be much lower than what you think you need.
> 
> For more info, make sure to check out [this post](https://github.com/brettwooldridge/HikariCP/wiki/About-Pool-Sizing)(and the video!). While the technology is different (postgres), a lot of the fundamentals are applicable for all systems, including ZIO.
> 
> EDIT: There is a fair bit of literature online on thread pool sizing / limiting parallelism, but you can use this formula as a starting point `Number of Cores * (1 + Blocking Coefficient)` where the Blocking Coefficient is the wait-to-CPU time ratio ([source](https://www.linkedin.com/pulse/demystifying-java-thread-pool-sizing-insights-optimal-rathnayake-rm47c))

```scala
  def run = {
    Ref.make(0).forEachZIO {
      perfomance => {
        val curTime = java.lang.System.currentTimeMillis()
        zStreamIp.mapZIOParUnordered(n) {
          ip =>
            for {
              ipaddr <- InetAddress.byName(ip.toString)
              r <- ZIO.blocking(ipaddr.isReachable(10000))
              _ <- perfomance.update(_ + 1)
              p <- perfomance.get
              _ <- Console.printLine(p * 1000 / (java.lang.System.currentTimeMillis() -  curTime))
              //_ <- if(r) Console.printLine(ip) else ZIO.none
            } yield ()
        }.runDrain
      }
    }
  }
```

the number of threads - 1500-2000 for n = Int.MaxValue, 20-200 for n = 32

am i wrong?

**Comment by myazinn** [comment]
> can lead to fork-kill loops where we kill fibers 'just before' we need them

@eyalfa it it's any helpful, I did try to play with a solution that avoids this issue. I tried to linearise killing/spawning fibers via yet another queue (see [here](https://github.com/myazinn/zio/pull/1/files)). And it did (kinda) work with relatively low performance overhead (depending on benchmark type, either 10% drop, or 25% drop with parallelism = 1, 15% drop with parallelism = 2, and no measurable difference with parallelism >= 4).
But when making stress tests I've noticed that it introduces another memory leak which I couldn't pinpoint 😞 Overall the solution does work, but from time to time (sometimes it takes minutes to run) it stars leaking quickly. TBH I'm still not sure if it's the issue with my solution or with something else inside ZIO, since it was extremely painful to minimise.
Also, as far as I remember, I came to a conclusion that I cannot just kill fibers wherever I want since it'll create a lot of pressure on readerCh (q1 queue), and potentially mess with it's internal state since it counts the amount of finished streams. I guess that can also be solved, but I couldn't do it quickly at that time

**Comment by kyri-petrou** [comment]
@myazinn could be very wrong, but that kind of behaviour sounds like something that could be explained by a concurrent update not done atomically, placing something into an invalid state.

~~Just a shot in the dark, but I think the current code uses `==` to check for when we've reached the max number of fibers. Maybe it needs to change to `>=` instead?~~

Edit: Actually ignore the second paragraph

**Comment by eyalfa** [comment]
@myazinn I'd appreciate if u can share the branch, I'd be happy to take a look.
@kyri-petrou , one thing I've noticed when playing with this (a bit), using threads-snapshot:
seems like the queue implementation is sub-optimal when 'lots' of fibers attempt to offer on a full queue, I was able to capture this stack trace:
```
ZScheduler-Worker-4" #25 [25347] daemon prio=5 os_prio=31 cpu=20973.27ms elapsed=76.41s tid=0x000000014680ac00 nid=25347 runnable  [0x000000031861e000]
   java.lang.Thread.State: RUNNABLE
	at scala.runtime.BoxesRunTime.equals2(BoxesRunTime.java:133)
	at scala.runtime.BoxesRunTime.equals(BoxesRunTime.java:119)
	at zio.internal.MutableConcurrentQueue.pollUpTo(MutableConcurrentQueue.scala:127)
	at zio.Queue$.zio$Queue$$unsafePollAll(Queue.scala:478)
	at zio.Queue$Strategy$BackPressure.unsafeOnQueueEmptySpace(Queue.scala:388)
	at zio.Queue$$anon$1.$anonfun$take$1(Queue.scala:257)
	at zio.Queue$$anon$1$$Lambda/0x0000007001199878.apply(Unknown Source)
	at zio.ZIO$.$anonfun$fiberIdWith$1(ZIO.scala:3266)
	at zio.ZIO$$$Lambda/0x0000007001155340.apply(Unknown Source)
	at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1055)
	at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1064)
```

seems like this fiber performs a `take` operation on a (very) full queue, trying to unblock a putter and ending up recycling the entire `putters` queue. since this queue is bounded (according to the `BackPressure` strategy in the stack trace) I deduce this is the second queue, looking at the code it seems `putters` should be using a `dequeue` like data structure instead of a queue.
* increasing the buffer size seems to prevent this issue and effectively cap the number of fibers, this is not really a viable solution as the 'required' buffer size is highly dependent on upstream vs downstream paces.

another bottleneck identified using threads-snapshot is the weak concurrent bag which performs a (long) gc when growing:

```
ZScheduler-Worker-1" #22 [28419] daemon prio=5 os_prio=31 cpu=12136.90ms elapsed=180.57s tid=0x000000013080d400 nid=28419 runnable  [0x000000017e732000]
   java.lang.Thread.State: RUNNABLE
	at java.util.Collection.removeIf(java.base@21.0.3/Collection.java:583)
	at zio.internal.WeakConcurrentBag.gc(WeakConcurrentBag.scala:80)
	at zio.internal.WeakConcurrentBag.add(WeakConcurrentBag.scala:66)
	at zio.internal.FiberScope$global$.add(FiberScope.scala:66)
	at zio.ZIO$unsafe$.makeChildFiber(ZIO.scala:2704)
	at zio.ZIO$unsafe$.fork(ZIO.scala:2665)
	at zio.ZIO.$anonfun$forkWithScopeOverride$2(ZIO.scala:827)
	at zio.ZIO$$Lambda/0x0000000801159af0.apply(Unknown Source)
	at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:964)
	at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1033)
...
```

**Comment by myazinn** [comment]
> but that kind of behaviour sounds like something that could be explained by a concurrent update not done atomically

@kyri-petrou that was my thought as well, but I couldn't find. TBH I still don't fully understand current solution (usages of `SingleProducerAsyncInput` blow my mind), so I had to give up

@eyalfa here's the branch rebased on latest master: https://github.com/myazinn/zio/tree/try-fixing-uncontrollable-grow-in-ZStream%23flatMapPar-1

**Comment by eyalfa** [comment]
> > but that kind of behaviour sounds like something that could be explained by a concurrent update not done atomically
> 
> @kyri-petrou that was my thought as well, but I couldn't find. TBH I still don't fully understand current solution (usages of `SingleProducerAsyncInput` blow my mind), so I had to give up

tldr: it's not really used, u can ignore it, it's there to convince the type system and the channel executor that the channel has its input 'sorted out'. if u take a close look at the channel's input type (both the 'outer' and inner channels) u'd see it's not `Nothing` since the channel API is not limited to streams. however both outer and inner channel are effectively streams, meaning connecting them to the queue reader has no effect.
> 
> @eyalfa here's the branch rebased on latest master: https://github.com/myazinn/zio/tree/try-fixing-uncontrollable-grow-in-ZStream%23flatMapPar-1

I'll hopefully have a look later today

**Comment by eyalfa** [comment]
@jgordijn did u make any progress profiling the memory leak?
did u try using the `kill -INFO` mechanism to get fiber dumps?
one more thing I started thinking about, is it possible your workload produces child fibers while consuming the streams from kafka? haven't confirmed it yet, but I suspect it's possible the old impl 'disposed' such fibers for you while the new one keeps them around

**Comment by jgordijn** [comment]
@eyalfa I’m away for a few weeks, so I cannot check it. I will get back here when I’m back in office. 

I will double check , but AFAIK there are no fibers created (by me) in the stream.

**Comment by eyalfa** [comment]
@kyri-petrou ,
seems like the revert went all way to series/2.0.
The stream operators were optimized in two 'steps', only the second step introduced the fibers pool which eligibly introduced the memory leak.

I suggest reapplying #8863 and #8819, both PRs kept the original design of fork per stream/effect while optimizing the implementation.

I think we still need a memory profiling test/bench in order to make sure we're still on the safe side, both or the old impl and for the new one.

regarding the leak in the 'pooling' implementation, seems like the leak is limited to unbounded parallelization use cases. 
I suspect some of them are leaky in the old implementation as well (fast upstream, slow consumer), 
some of them are actually leaky in the new implementation: when handling bursts,
and the interesting case (to me) is the kafka stream scenario where the the number of streams is quite small but they live (almost) forever: I'm still waiting for a heap and thread dumps from @jgordijn so very hard to tell what's going on there.

another aspect I suspected was accumulation of child fibers or FiberRefs updates in the long living 'pooled' fibers, guess we'll never really look into it...

**Comment by kyri-petrou** [comment]
@eyalfa we can discuss this more on Discord, but since we're unsure on where exactly the memory leaks are coming from, I think it's better to play it _very_ safe for the time being and revert all of the changes.

Once we have a new release (shortly) can then revisit those PRs are apply the optimizations separately, although this time we'll need a much more robust test suite to accomodate the changes, and the benchmarks should also be run with unbounded parallelism.

**Comment by eyalfa** [comment]
sure, safety first 😎 
let's keep this issue open or alternatively open a new one in order to track the required tests for memory consumption.
I'd still like to see the fibers and memory dumps from @jgordijn, and a concrete reproducer that doesn't leak in old impl as well (unlike unbounded parallelism with a fast producer and slow consumer).

**Comment by jgordijn** [comment]
I can confirm that the memory behavior is back to normal with the latest release.

**Comment by kyri-petrou** [comment]
Thanks for confirming @jgordijn!

@regiskuckaertz I'm going to close this issue; if you want to collect the $150 bounty (since you made the fix) perhaps reach out to [this Discord channel](https://discord.com/channels/1040418099612098610/1040423036035268700) and ask if they can manually assign it to you

**Comment by regiskuckaertz** [comment]
I'm only here for fame and glory
